clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java

```diff
@@ -892,34 +903,118 @@ public ListOffsetResult() {
+                        ApiKeys.OFFSET_FOR_LEADER_EPOCH.latestVersion(), dataMap);
+                sendOffsetForLeaderEpochRequest(node, builder);
+            });
+        }
+
         Map<Node, FetchSessionHandler.FetchRequestData> reqs = new LinkedHashMap<>();
         for (Map.Entry<Node, FetchSessionHandler.Builder> entry : fetchable.entrySet()) {
             reqs.put(entry.getKey(), entry.getValue().build());
         }
         return reqs;
     }
 
+    private <T> Map<Node, Map<TopicPartition, T>> regroupPartitionMapByNode(Map<TopicPartition, T> partitionMap) {
+        return partitionMap.entrySet()
+                .stream()
+                .collect(Collectors.groupingBy(entry -> metadata.fetch().leaderFor(entry.getKey()),
+                        Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue)));
+    }
+
+    private RequestFuture<OffsetsForLeaderEpochResponse> sendOffsetForLeaderEpochRequest(
+            Node node, OffsetsForLeaderEpochRequest.Builder builder) {
+
+        return client.send(node, builder).compose(new RequestFutureAdapter<ClientResponse, OffsetsForLeaderEpochResponse>() {
+            @Override
+            public void onSuccess(ClientResponse response, RequestFuture<OffsetsForLeaderEpochResponse> future) {
+                OffsetsForLeaderEpochResponse ofler = (OffsetsForLeaderEpochResponse) response.responseBody();
+                log.trace("Received OffsetsForLeaderEpochResponse {} from broker {}", ofler, node);
+
+                Map<TopicPartition, OffsetAndMetadata> truncationWithoutResetPolicy = new HashMap<>();
+                ofler.responses().forEach((respTopicPartition, respEndOffset) -> {
+                    // For each OffsetsForLeader response, check for truncation. This occurs iff the epoch has changed
+                    // and the end offset is lower than our current offset
+                    ConsumerMetadata.LeaderAndEpoch currentLeaderAndEpoch = metadata.leaderAndEpoch(respTopicPartition);
+                    ConsumerMetadata.LeaderAndEpoch leaderAndEpochForEndOffset = new ConsumerMetadata.LeaderAndEpoch(
+                            Node.noNode(), Optional.of(respEndOffset.leaderEpoch()));
+
+                    if (currentLeaderAndEpoch.isObsoletedBy(leaderAndEpochForEndOffset)) {
+                        // If we have a new epoch, check if our offset is ahead of the new end offset
+                        SubscriptionState.FetchPosition currentPosition = subscriptions.position(respTopicPartition);
+                        if (respEndOffset.endOffset() < currentPosition.offset) {
+                            // Try to reset the offset
+                            if (subscriptions.hasDefaultOffsetResetPolicy()) {
+                                log.info("Truncation detected for partition {}, resetting offset", respTopicPartition);
+                                subscriptions.requestOffsetReset(respTopicPartition);
```

Comment:
I think the point of this is to set the fetch offset to the offset and epoch from the response.

Reply:
Ok, so we detect truncation (leader's end offset is lower than the client's next fetch offset). Then we need to reposition ourselves. Do we consider this a reset of any kind? IOW, should the reset policy even be considered here?

Maybe we simply move to the new end-offset and carry on. But in that case, we'll never hit the LogTruncationException error condition.

Reply:
This logic implemented now is to reset to the end offset from the response iff the consumer has a default reset policy defined (doesn't matter which one). If we detect truncation and no offset reset policy is defined, we raise an error.